packages/iov-core/src/transactionencoder.ts

```diff
@@ -0,0 +1,101 @@
+import { Encoding } from "@iov/encoding";
+import { JsonCompatibleValue } from "@iov/jsonrpc";
+
+/**
+ * Encodes non-circular JavaScript objects and primitives into JSON.
+ * Used for encoding/decoding transactions but works for kind of data consisting of the supported types.
+ *
+ * Supported types:
+ * - boolean
+ * - number
+ * - null
+ * - object
+ * - Array
+ * - string
+ * - Uint8Array
+ */
+export class TransactionEncoder {
+  public static toJson(data: unknown): JsonCompatibleValue {
+    if (typeof data === "number" || typeof data === "boolean") {
+      return data;
+    }
+
+    if (data === null) {
+      return null;
+    }
+
+    if (typeof data === "string") {
+      return `string:${data}`;
+    }
+
+    if (data instanceof Uint8Array) {
+      return `bytes:${Encoding.toHex(data)}`;
+    }
+
+    if (Array.isArray(data)) {
+      return data.map(TransactionEncoder.toJson);
+    }
+
+    // Exclude special kind of objects like Array, Date or Uint8Array
+    // Object.prototype.toString() returns a specified value:
+    // http://www.ecma-international.org/ecma-262/7.0/index.html#sec-object.prototype.tostring
+    if (
+      typeof data === "object" &&
+      data !== null &&
```

Comment:
`null` check not strictly necessary because of above `return`.

Reply:
Right. Unfortunately TypeScript's type inference is not great enough to know that. `typeof data === "object"` leads to `object | null` in this block, since `typeof null` is `object`.